### ISSUE TITLE
Deprecation process: refine it to handle deprecated functiosn and class methods

### ISSUE THREAD

**Issue by josephsl** [nvaccess/nvda#15668]
### Is your feature request related to a problem? Please describe.
At the moment the getattr method of deprecating symbols works well with constants. However, when functions are deprecated, using getattr method causes Python to raise attribute error.

### Describe the solution you'd like
Let us refine the getattr deprecation process to handle deprecated functions and class methods.

### Describe alternatives you've considered
Allow deprecated functions to be marked via conditional call to NVDAState._allowDeprecatedAPI() function like so:

if NVDAState._allowDeprecatedAPI():
    def deprecatedFunction(): pass

### Additional context
See #15666 for detailed discussion and motivation.

Thanks.

**Comment by CyrilleB79** [comment]
Maybe a decorator to deprecate a function ould be suitable for this.

**Comment by lukaszgo1** [comment]
I agree that we don't have a good way to deprecate class / static / instance methods at present. This should be done either with a decorator as suggested above or with a metaclass. However the initial motivation for this issue seems to be about module level functions, which can be deprecated using `__getattr__` easily enough.
Note that doing this with `__getattr__` as opposed to a decorator, or showing a warning inside the function means that the deprecated code is not shown when auto completing in the Python console, nor in IDE's making it less likely someone uses it in core or add-ons.

> At the moment the getattr method of deprecating symbols works well with constants. However, when functions are deprecated, using getattr method causes Python to raise attribute error.


Would you be able to show the code you have written which causes this? In case of `winVersion.isFullScreenMagnificationAvailable` it can be deprecated using `__getatrr__` in two ways:
1. Assuming we do not  care about its doc string adding the following at the bottom of `winVersion.__getattr__`, just before raising of `AttributeError` works:
```python
if attrName == "isFullScreenMagnificationAvailable" and NVDAState._allowDeprecatedAPI():
		log.debugWarning(
			"Deprecated function called: winVersion.isFullScreenMagnificationAvailable, "
			"use visionEnhancementProviders.screenCurtain.ScreenCurtainProvider.canStart instead.",
			stack_info=True
		)
		return lambda: True

```
2. If the doc string has to be kept, and the function should look as a real function i.e. being a `lambda` is problematic for whatever reason,  you can just move content of the `isFullScreenMagnificationAvailable` inside `__getattr__` as a inner function, and return it from the module level `__getattr__`.

I'm really curious what approach you have tried.

**Comment by josephsl** [comment]
Hi,

Initially, I tried putting the contents of the deprecated function inside getattr and removed the deprecated function from the module, which did work when called as a symbol but failed when called as a function. I then looked at gui.quit where it does log a deprecation notice, and that function itself is defined if deprecation API is allowed. I ended up using the second method without realizing that getattr method is the way to go when presenting deprecation notices (at least it shows how I haven't kept up with project docs lately).

Decorating deprecated functions is one way to do so, with the decorator argument being a notice about replacements. But then it raises the question of how to extend this to symbols and class attributes.

Thanks.

**Comment by seanbudd** [comment]
I don't think we should be using that syntax generally.

1. >Note that doing this with __getattr__ as opposed to a decorator, or showing a warning inside the function means that the deprecated code is not shown when auto completing in the Python console, nor in IDE's making it less likely someone uses it in core or add-ons.

2. Also using `__getattr__`  makes the logging only occur on import, which is much less noisy than on function call

**Comment by XLTechie** [comment]
Could a decorator be appropriate for this?

**Comment by seanbudd** [comment]
I think this has been handled by #18809
